internal/application/manager.go

```diff
@@ -191,9 +224,148 @@ func (m *dataManager) RecordingStatus() *dtos.RecordStatus {
 
 // StartReplay starts a replay session based on the values in the request
 // An error is returned if the request data is incomplete or a record or replay session is currently running.
-func (m *dataManager) StartReplay(request *dtos.ReplayRequest) error {
-	//TODO implement me using TDD
-	return errors.New("not implemented")
+func (m *dataManager) StartReplay(request dtos.ReplayRequest) error {
+	m.recordingMutex.Lock()
+	defer m.recordingMutex.Unlock()
+
+	if m.recordingStartedAt != nil {
+		return recordingInProgressError
+	}
+
+	if m.replayStartedAt != nil {
+		return replayInProgressError
+	}
+
+	if m.recordedData == nil {
+		return noRecordedData
+	}
+
+	if request.ReplayRate <= 0 {
+		return invalidReplayRate
+	}
+
+	if request.RepeatCount < 0 {
+		return invalidReplayCount
+	}
+
+	now := time.Now()
+	m.replayStartedAt = &now
+	m.replayedDuration = 0
+	m.replayedEventCount = 0
+	m.replayedRepeatCount = 0
+	m.replayError = nil
+	m.replayContext, m.replayCancelFunc = context.WithCancel(context.Background())
+
+	go m.replayRecordedEvents(request)
+
+	return nil
+}
+
+func (m *dataManager) replayRecordedEvents(request dtos.ReplayRequest) {
+	var previousEventTime int64
+	firstEvent := true
```

Comment:
should we lock from the get-go if this app is running in multiple threading?

Reply:
Nope, since that will block get Replay Status and Cancel Status from accessing the needed fields

Reply:
i see